src/expr/node.h

```diff
@@ -1331,6 +1360,117 @@ TypeNode NodeTemplate<ref_count>::getType(bool check) const
   return NodeManager::currentNM()->getType(*this, check);
 }
 
+template <bool ref_count>
+inline Node NodeTemplate<ref_count>::substituteCaptureAvoiding(
+    Node node, Node replacement) const
+{
+  if (node == *this)
+  {
+    return replacement;
+  }
+  std::vector<Node> source;
+  std::vector<Node> dest;
+  source.push_back(node);
+  dest.push_back(replacement);
+  std::unordered_map<Node, Node, NodeHashFunction> cache;
+  return substituteCaptureAvoiding(source, dest, cache);
+}
+
+template <bool ref_count>
+Node NodeTemplate<ref_count>::substituteCaptureAvoiding(
+    Node node,
+    Node replacement,
+    std::unordered_map<Node, Node, NodeHashFunction>& cache) const
+{
+  std::vector<Node> source;
+  std::vector<Node> dest;
+  source.push_back(node);
+  dest.push_back(replacement);
+  return substituteCaptureAvoiding(source, dest, cache);
+}
+
+template <bool ref_count>
+inline Node NodeTemplate<ref_count>::substituteCaptureAvoiding(
+    std::vector<Node> source, std::vector<Node> dest) const
+{
+  std::unordered_map<Node, Node, NodeHashFunction> cache;
+  return substituteCaptureAvoiding(source, dest, cache);
+}
+
+template <bool ref_count>
+Node NodeTemplate<ref_count>::substituteCaptureAvoiding(
+    std::vector<Node> source,
+    std::vector<Node> dest,
+    std::unordered_map<Node, Node, NodeHashFunction>& cache) const
+{
+  // in cache?
+  typename std::unordered_map<Node, Node, NodeHashFunction>::const_iterator i =
+      cache.find(*this);
+  if (i != cache.end())
+  {
+    return (*i).second;
+  }
+
+  // otherwise compute
+  Assert(source.size() == dest.size(),
+         "Substitution domain and range must be equal size");
+
+  auto it = std::find(source.begin(), source.end(), (*this));
+  if (it != source.end())
+  {
+    Assert(std::distance(source.begin(), it) >= 0
+           && std::distance(source.begin(), it) < dest.size());
+    Node n = dest[std::distance(source.begin(), it)];
+    cache[*this] = n;
+    return n;
+  }
+  if (getNumChildren() == 0)
+  {
+    cache[*this] = *this;
+    return *this;
+  }
+  bool binder = isClosure();
+  // if binder, rename variables to avoid capture
+  if (binder)
+  {
+    std::vector<Node> vars;
+    std::vector<Node> renames;
+
+    NodeManager* nm = NodeManager::currentNM();
+    for (const Node& v : (*this)[0])
+    {
+      vars.push_back(v);
+      renames.push_back(nm->mkBoundVar(v.getType()));
+    }
+    // have new vars -> renames subs in the beginning of current sub
+    source.insert(source.begin(), vars.begin(), vars.end());
+    dest.insert(dest.begin(), renames.begin(), renames.end());
+  }
+  NodeBuilder<> nb(getKind());
+  if (getMetaKind() == kind::metakind::PARAMETERIZED)
+  {
+    // push the operator
+    nb << getOperator().substituteCaptureAvoiding(source, dest, cache);
+  }
+  for (const_iterator i = begin(), iend = end(); i != iend; ++i)
+  {
+    nb << (*i).substituteCaptureAvoiding(source, dest, cache);
```

Comment:
We should make this function non-recursive. Right now, we are copying the `source` and the `dest` vectors every time we go to a child node and the nesting of nodes can get very deep.

Reply:
OK, will do